src/sctools/bam.py

```diff
@@ -1,55 +1,107 @@
+    ----------
+    alignment_file : str
+        sam or bam file
+    open_mode : {'r', 'rb', None}, optional
+        open mode for pysam.AlignmentFile. 'r' indicates a sam file, 'rb' indicates a bam file,
+        and None attempts to autodetect based on the file suffix (Default = None)
+
+    Methods
+    -------
+    indices_by_chromosome
+        returns indices to line numbers containing the requested number of reads for a specified
+        chromosome
+
+    Notes
+    -----
+    samtools is a good general-purpose tool for that is capable of most subsampling tasks. It is a
+    good idea to check the samtools documentation when approaching these types of tasks.
+
+    References
+    ----------
+    samtools documentation : http://www.htslib.org/doc/samtools.html
 
-    :method indices_by_chromosome: extract reads specific to a chromosome
     """
 
-    def __init__(self, alignment_file, open_mode=None):
-        """
-        :param str alignment_file: sam or bam file.
-        :param str open_mode: optional, mode to read file. Will be autodetected by file type if
-          the file contains the correct suffix for its type.
-        """
-
+    def __init__(self, alignment_file: str, open_mode: str=None):
         if open_mode is None:
             if alignment_file.endswith('.bam'):
                 open_mode = 'rb'
             elif alignment_file.endswith('.sam'):
                 open_mode = 'r'
             else:
-                raise ValueError('could not autodetect file type for alignment_file %s '
-                                 '(detectible suffixes: .sam, .bam)' % alignment_file)
-        self._file = alignment_file
-        self._open_mode = open_mode
-
-    def indices_by_chromosome(self, n_specific, chromosome, include_other=0):
-        """Return the list of first n_specific indices of reads aligned to selected chromosome.
-
-        If desired, will also return non-specific indices in a second list (can serve as negative
-        control reads).
-
-        :param int n_specific: number of aligned reads to return indices for
-        :param str chromosome: only reads from this chromosome are considered valid
-        :param int include_other: optional, (default=0), the number of reads to include that are
-          NOT aligned to chromosome (could be aligned or unaligned read)
-        :return [int]: list of indices to reads aligning to chromosome
-        :return [int]: list of indices to reads NOT aligning to chromosome, only returned if
-          include_other is not 0.
+                raise ValueError(
+                    'could not autodetect file type for alignment_file %s (detectible suffixes: '
+                    '.sam, .bam)' % alignment_file
+                )
+        self._file: str = alignment_file
+        self._open_mode: str = open_mode
+
+    # todo figure out how to generate optional output type hints
+    def indices_by_chromosome(self, n_specific: int, chromosome: str, include_other: int=0):
```

Comment:
The inputs are using PEP484, I'd also make the return values comply with PEP484, i.e. `def indices_by_chromosome(self, n_specific: int, chromosome: str, include_other: int=0) -> ???` for this case. Although Numpy docstring force you to duplicate the return type in the docstring, I think adding `->` is better, from the "self-documented code" perspective. 

Besides, for the todo, I guess:
```python
from typing import Union

f(x) -> Union[int, None]:
```
could be a feasible solution here?

Reply:
Ah, after going through the implementation, I feel the annotations would be:

```python
from typing import Union, Tuple
def indices_by_chromosome(self, n_specific: int, chromosome: str, include_other: int=0) -> Union[Tuple[List[int], List[int]], List[int]]:
```
This seems to be more complicated now, a pain of Typing :(

Reply:
We agree on this implementation. Added it.